rpctest/rpc_harness.go

```diff
@@ -100,6 +106,20 @@ type Harness struct {
 	sync.Mutex
 }
 
+// NewWithDCRD creates and initializes new instance of the rpc test harness.
+// The dcrd node located at pathToDCRD is used if supplied, otherwise one is
+// created in a temporary directory. Optionally, websocket handlers and a
+// specified configuration may be passed. In the case that a nil config is
+// passed, a default configuration will be used.
+//
+// NOTE: This function is safe for concurrent access.
+func NewWithDCRD(t *testing.T, activeNet *chaincfg.Params, handlers *rpcclient.NotificationHandlers, extraArgs []string, pathToEXE string) (*Harness, error) {
+	pathToDCRDMtx.Lock()
+	pathToDCRD = pathToEXE
+	pathToDCRDMtx.Unlock()
```

Comment:
This has now become a side effect and makes New/NewWithDCRD non idempotent and impossible to reason about in concurrent settings.

Say you have the following calls (pseudocode):

```go
go NewWithDcrd(..., "/bin/first-dcrd") // #1
go New() // #2
go NewWithDcrd(..., "/bin/second-dcrd") // #3
```
Which binary is `#2` going to use? It could use either `/bin/first-dcrd`, `/bin/second-dcrd` or even a newly built one (if its goroutine happens to run first). The API now doesn't really clearly express that calls for `NewWithDcrd()` affect calls for `New()` and that such calls need to be synchronized.

So IMO, _either_ New() should always build the node (and thus external callers are forced to _always_ use NewWithDCRD, _or_ you should expose a `SetDCRDPath()`, making `New()` always use the global path (and making explicit for users that calls for new nodes need to be synchronized externally if they are to use different binaries).

Reply:
Does this make it better?